src/widgets/FreeformInputWidget.tsx

```diff
@@ -0,0 +1,289 @@
+  name: string
+  label: string
+  help?: string | null
+  required: boolean
+  details: FreeformInputWidgetDetails
+}
+
+interface FreeformInputWidgetProps {
+  configuration: FreeformInputWidgetConfiguration
+  /**
+   * Whether the underlying fieldset should be functionally and visually disabled.
+   */
+  fieldsetDisabled?: boolean
+  /**
+   * Whether to hide the widget label from ARIA.
+   */
+  labelAriaHidden?: boolean
+  /**
+   * When true, bypass the required attribute if all options are made unavailable by constraints.
+   */
+  bypassRequiredForConstraints?: boolean
+}
+
+const FreeformInputWidget = ({
+  configuration,
+  labelAriaHidden = true,
+  fieldsetDisabled
+}: FreeformInputWidgetProps) => {
+  const [value, setValue] = useState<string>()
+
+  const persistedSelection = useReadLocalStorage<{
+    dataset: { id: string }
+    inputs: { [k: string]: string | number }
+  }>('formSelection')
+
+  const { type, help, label, name, required, details } = configuration
+
+  /**
+   * Cache persisted selection, so we don't need to pass it as an effect dependency.
+   */
+  const persistedSelectionRef = useRef(persistedSelection)
+
+  useEffect(() => {
+    const getInitialSelection = (): string => {
+      if (
+        persistedSelectionRef.current &&
+        'inputs' in persistedSelectionRef.current
+      ) {
+        if (
+          typeof persistedSelectionRef.current.inputs[name] === 'string' &&
+          details.dtype === 'string'
+        ) {
+          return persistedSelectionRef.current.inputs[name] as string
+        } else if (
+          typeof persistedSelectionRef.current.inputs[name] === 'number' &&
+          (details.dtype === 'float' || details.dtype === 'int')
+        ) {
+          return persistedSelectionRef.current.inputs[name].toString()
+        }
+      }
+
+      return (details?.default ?? '').toString()
+    }
+
+    setValue(getInitialSelection())
+  }, [name, details])
+
+  if (!configuration) return null
+
+  if (type !== 'FreeformInputWidget') return null
+
+  const inputRef = useRef<HTMLInputElement>(null)
+  const fieldSetRef = useRef<HTMLFieldSetElement>(null)
+
+  const {
+    dtype,
+    default: defaultValue,
+    comment
+  } = configuration.details as FreeformInputWidgetDetails
+
+  const { inputType, otherProps } = useInputType(dtype)
+
+  const initialValue = value ?? defaultValue ?? ''
+
+  return (
+    <Widget data-stylizable='widget freeform-input-widget'>
+      <WidgetHeader>
+        <WidgetTitle
+          data-stylizable='widget-title'
+          htmlFor={name}
+          aria-hidden={labelAriaHidden}
+        >
+          {label}
+        </WidgetTitle>
+        <WidgetTooltip
+          helpText={help || null}
+          triggerAriaLabel={`Get help about ${label}`}
+        />
+      </WidgetHeader>
+      <ReservedSpace data-stylizable='widget freeform-input reserved-error-space'>
+        {required && value?.length ? (
+          <Error>The field is required.</Error>
```

Comment:
I think it should be "This field is required"

Reply:
Also, it seems that the message does not disappear even if the value is present.

Reply:
Fixed!